### PR TITLE
add aoc day2 part1 using udfs

### PR DESCRIPTION
**Pull request by guo543**

### CODE REVIEW

File: test/aoc.test.js [hunk 1]
```diff
@@ -75,6 +75,61 @@ let res = func({input, udf})
...
+    },
+    extractId: x => Number(x.substring(5)),
+    all: x => x.every(v => v) ? 1 : 0,
+    check: x => {
+      // x is expected to be in the form of '<number> <color>'
+      let s = x.trim()
+      let split = s.split(" ")
+      let n = Number(split[0])
+      return n <= bag[split[1]]
+    }
```
Comment (TiarkRompf) [inline test/aoc.test.js:108]: Let's break these down further. UDFs should be elementary, i.e., do one thing only that's presently not supported by Rhyme. Here's we have quite a bit of application logic in `extractId`, `all`, and especially `check`.

File: test/aoc.test.js [hunk 2]
```diff
@@ -75,6 +75,61 @@ let res = func({input, udf})
...
+    .map("udf.splitColon").get("*part")
+
+  let id = api.apply("udf.extractId", api.first(lines))
+  let isPossible = api.array(api.apply("udf.check", api.get(api.apply("udf.scan", api.last(lines)), "*match")))
+
+  let lineRes = api.times(id, api.apply("udf.all", isPossible))
+
+  let query = 
+    pipe(lineRes).group("*line").get("*").sum()
+
```
Comment (TiarkRompf) [inline test/aoc.test.js:125]: Don't we need to split the input by ';' and ',' at some point?

**Review by TiarkRompf — CHANGES_REQUESTED** [review]

File: test/aoc.test.js (hunk 2 above)
Comment (guo543) [reply test/aoc.test.js:125]: I followed the approach of the jq solution where they used scan to match the pattern <num> <color>. We can try to change that as well.

Comment (TiarkRompf) [reply test/aoc.test.js:125]: Yeah, reading the problem description more closely I think the turns-per-game grouping is irrelevant for the result, but I'd still prefer to split the input "properly"

Do you have part 2 working as well?

Comment (guo543) [reply test/aoc.test.js:125]: I'm working on part2. I think for part2 there might also be some application logic in udfs. I'll change part1 first and try to make udfs elementary.

Comment (TiarkRompf) [reply test/aoc.test.js:125]: OK. I found the Scala solutions quite readable as well: https://scalacenter.github.io/scala-advent-of-code/2023/puzzles/day02

**Pull request closed without merging**